test/serve/build_delay_test.dart

```diff
@@ -72,52 +54,107 @@ main() {
     await endPubServe();
   });
 
-  test("continual fast edits won't cause multiple builds", () async {
-    var pubServeProcess =
-        await pubServe(forcePoll: false, args: ['--build-delay', '100']);
-    expect(pubServeProcess.stdout,
-        emitsThrough(contains('Build completed successfully')));
-    await requestShouldSucceed("packages/myapp/lib.dart", "foo() => 'foo';");
-
-    for (var i = 0; i < 10; i++) {
-      writeTextFile(libFilePath, "foo() => '$i';");
-      await new Future.delayed(new Duration(milliseconds: 50));
+  group('unit tests', () {
+    Barback barback;
+    List<BuildResult> barbackResults;
+    AssetEnvironment environment;
+    final libAssetId = new AssetId('myapp', 'lib/lib.dart');
+    _MockDirectoryWatcher libWatcher;
+    _MockWatcherType watcherType;
+
+    setUp(() async {
+      var entrypoint =
+          new Entrypoint(p.join(d.sandbox, appPath), new SystemCache());
+      watcherType = new _MockWatcherType();
+      environment = await AssetEnvironment.create(entrypoint, BarbackMode.DEBUG,
+          watcherType: watcherType,
+          buildDelay: new Duration(milliseconds: 50),
+          compiler: Compiler.none);
+      barback = environment.barback;
+      libWatcher = watcherType.watchers[p.join(d.sandbox, appPath, 'lib')];
+      // Collect build results.
+      barbackResults = <BuildResult>[];
+      barback.results.listen(barbackResults.add);
+    });
+
+    tearDown(() async {
+      await environment.cleanUp();
+    });
+
+    // Attempts to wait for all pending barback builds.
+    Future _waitForBarback() async {
+      // First, wait for the next build to complete.
+      await barback.results.first;
+      // Then wait for all assets, which should capture additional builds if
+      // they occur.
+      await barback.getAllAssets();
+      // Give the stream a chance to deliver a new build if one did occur.
+      await new Future(() {});
     }
 
-    // Should only see one build.
-    expect(pubServeProcess.stdout,
-        emitsThrough(contains('Build completed successfully')));
-    expect(pubServeProcess.stdout, neverEmits('Build completed successfully'));
+    test("continual fast edits won't cause multiple builds", () async {
+      expect(await (await barback.getAssetById(libAssetId)).readAsString(),
+          "foo() => 'foo';");
+
+      for (var i = 0; i < 10; i++) {
+        writeTextFile(libFilePath, "foo() => '$i';");
+        libWatcher.addEvent(new WatchEvent(ChangeType.MODIFY, libFilePath));
+        await new Future.delayed(new Duration(milliseconds: 15));
```

Comment:
Technically this could still be flaky just due to async scheduling, so I would like to avoid it. I doubt we would ever get a flake as it is now (and its still much faster than the integration tests).